v2/apiserver/internal/core/events.go

```diff
@@ -537,33 +537,41 @@ func (e *eventsService) CancelMany(
 		)
 	}
 
-	events, err := e.eventsStore.CancelMany(ctx, selector)
+	eventCh, affectedCount, err := e.eventsStore.CancelMany(ctx, selector)
 	if err != nil {
 		return result, errors.Wrap(err, "error canceling events in store")
 	}
-
-	result.Count = int64(len(events.Items))
-
-	// TODO: This could take a while, so we don't do it synchronously. But what if
-	// the process dies while this is in-progress? Can we find a quicker, more
-	// efficient way to do this?
-	go func() {
-		for _, event := range events.Items {
-			if err := e.substrate.DeleteWorkerAndJobs(
-				context.Background(), // Deliberately not using request context
-				project,
-				event,
-			); err != nil {
-				log.Println(
-					errors.Wrapf(
-						err,
-						"error deleting event %q worker and jobs from the substrate",
-						event.ID,
-					),
-				)
+	result.Count = affectedCount
+
+	// Fan out to a finite number of goroutines to handle cleanup duties
+	concurrency := 10
+	if result.Count < int64(concurrency) {
+		concurrency = int(result.Count)
+	}
+	for i := 0; i < concurrency; i++ {
+		go func() {
+			for {
+				select {
+				case event, ok := <-eventCh:
+					if ok {
```

Comment:
Nit: If we lead with the following, everything that follows it is indented one level less, which is generally considered to reduce the complexity.

```
if !ok {
  return
}
```